src/aspire/source/__init__.py

```diff
@@ -117,32 +118,28 @@ def states(self, values):
         return self.set_metadata('_rlnClassNumber', values)
 
     @property
-    def filters(self):
-        return self.get_metadata('__filter')
-
-    @filters.setter
-    def filters(self, values):
-        self.set_metadata('__filter', values)
-        if values is None:
-            new_values = np.nan
+    def filter_indices(self):
+        return self.get_metadata('__filter_indices')
+
+    @filter_indices.setter
+    def filter_indices(self, indices):
+        # create metadata of filters for all images
+        if indices is None:
+            filter_values = np.nan
         else:
-            new_values = np.array([(
-                getattr(f, 'voltage', np.nan),
-                getattr(f, 'defocus_u', np.nan),
-                getattr(f, 'defocus_v', np.nan),
-                getattr(f, 'defocus_ang', np.nan),
-                getattr(f, 'Cs', np.nan),
-                getattr(f, 'alpha', np.nan)
-            ) for f in values])
+            attribute_list = ('voltage', 'defocus_u', 'defocus_v',
```

Comment:
I don't understand what's going on here. Why can't we just set the `filter_indices` metadata?

Reply:
The Relion Class has metadata for each image created from the star file in additional the unique filters. To make consistent between Simulation and Relion class, we would like to using the same structure of metadata so that It can be saved to a star file in the similar way. When we update the `filter_indices`, the relation information on the metadata table need to be updated too. So when we save metadata, it will have the updated information. If we don't do it here, we will do it before save the star file.

Reply:
Ok. This redundancy is problematic. We should be storing the CTF information in one place: the filter object. When we save to a STAR file, we should extract that information from the filter object (if it exists and is of the proper type). Having the same information stored in multiple places can cause problems down the road.

This is not something we should fix now, but let's make an issue to handle this.

Reply:
Opened a new issue #301.